java-checks/src/main/java/org/sonar/java/checks/SpecializedFunctionalInterfacesCheck.java

```diff
@@ -0,0 +1,239 @@
+        .map(interf -> new JavaFileScannerContext.Location("Interface Not Specialized", interf.classInterface))
+        .collect(Collectors.toList());
+      if (reportTreeAndStringInterfaces.size() > 1) {
+        reportIssue(((ClassTree) classInterfaces.parent()).simpleName(),
+          REPORT_STRING + "s '" + reportTreeAndStringInterfaces.stream().map(x -> x.reportString).collect(Collectors.joining(", ")) + "'", flow, null);
+      } else {
+        reportIssue(((ClassTree) classInterfaces.parent()).simpleName(), REPORT_STRING + " '" + reportTreeAndStringInterfaces.get(0).reportString + "'", flow, null);
+      }
+    }
+  }
+
+  public void checkVariableTypeAndInitializer(VariableTree variableTree) {
+    TypeTree interfaceType = variableTree.type();
+    ExpressionTree initializer = variableTree.initializer();
+    if (initializer != null && (initializer.is(Tree.Kind.LAMBDA_EXPRESSION) || isAnonymousClass(initializer))) {
+      matchFunctionalInterface(interfaceType)
+        .ifPresent(treeAndStringInterface -> reportIssue(treeAndStringInterface.classInterface, REPORT_STRING + " '" + treeAndStringInterface.reportString + "'"));
+    }
+  }
+
+  private static boolean isAnonymousClass(ExpressionTree initializeTree) {
+    return initializeTree.is(Tree.Kind.NEW_CLASS) && ((NewClassTree) initializeTree).classBody() != null;
+  }
+
+  private static Optional<InterfaceTreeAndStringPairReport> matchFunctionalInterface(TypeTree interfaceType) {
+    Optional<TypeArguments> args = interfaceParameters(interfaceType);
+    if (args.isPresent()) {
+      TypeArguments arguments = args.get();
+      Optional<String> reportString;
+      switch (interfaceType.symbolType().fullyQualifiedName()) {
+        case "java.util.function.Function":
+          reportString = handleFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiFunction":
+          reportString = handleBiFunctionInterface(arguments);
+          break;
+        case "java.util.function.BiConsumer":
+          reportString = handleBiConsumerInterface(arguments);
+          break;
+        case "java.util.function.Supplier":
+        case "java.util.function.Consumer":
+        case "java.util.function.Predicate":
+        case "java.util.function.UnaryOperator":
+        case "java.util.function.BinaryOperator":
+          reportString = handleOneParameterInterface(arguments).map(s -> s + interfaceType.symbolType().name());
+          break;
+        default:
+          return Optional.empty();
+      }
+      if (reportString.isPresent()) {
+        return Optional.of(new InterfaceTreeAndStringPairReport(reportString.get(), interfaceType));
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleOneParameterInterface(TypeArguments args) {
+    String reportString = new ParameterTypeNameAndTreeType(args.get(0)).paramTypeName;
+    if (reportString != null) {
```

Comment:
This whole test can be replaced by `return Optional.ofNullable(reportString);`.
It will automatically create an empty optional if its null.